tests/c_api_tests/test_cpp_api_serialization.cpp

```diff
@@ -54,12 +54,12 @@ TEST_CASE("C++ API Serialization and Deserialization") {
     std::vector<Idx> const total_elements = {1, 2};
 
     SUBCASE("Serializer") {
-        DatasetConst dataset{"input", is_batch, batch_size};
+        DatasetConst const dataset{"input", is_batch, batch_size};
```

Comment:
Thanks again! Same for all below.

Reply:
I actually see something weird here. The `add_buffer`/`add_attribute_buffer` functions in the CPP wrapper of const and mutable dataset are `const` functions. This is for me very confusing.

We are literally mutating the layout of the dataset, why are they const functions? In my opinion those functions should be declared as non-const, and this const specifier should be gone.

Reply:
They can be (and are) const-qualified because the wrapper object `Dataset*` itself does not mutate, since it only holds smart pointers to the core objects that contain the actual dataset (which does mutate). However, following the logic and intention, probably they should not be `const` functions. Nevertheless, since the `const` qualifier here is used by the compiler for optimizations, I wonder if we should still keep them. @Jerry-Jinfeng-Guo any thoughts on this?

Reply:
I was indeed simply following what was implemented in the core. I see this as extra encapsulation, not that I am in favor. If needed be, I could also apply the changes (which can be numerous).

Reply:
I do think we should remove the `const` function for this. `const` should be also expressing intentions. Like for `std::vector`, which also only stores a pointer to the actual array buffer. What if they allow this? It is theoretically possible but extremely confusing.

```c++
std::vector<int> const data{1, 2, 3, 4, 5};
data[2] = 100;
```

So yes, my proposal is to remove the const function and adjust all the tests. Instead of doing other way around.

Reply:
> `const` should be also expressing intentions

Strongly for this. Changes incoming.